x-pack/plugins/apm/public/components/shared/EnvironmentFilter/index.tsx

```diff
@@ -0,0 +1,123 @@
+// @ts-ignore
+import { EuiSuperSelect, EuiFlexGroup, EuiFlexItem } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import React, { useState, useEffect } from 'react';
+import styled from 'styled-components';
+import { fontSizes, units, px } from '../../../style/variables';
+import { useFetcher } from '../../../hooks/useFetcher';
+import { loadServiceEnvironments } from '../../../services/rest/apm/services';
+import { useLocation } from '../../../hooks/useLocation';
+import { useUiFilters } from '../../../hooks/useUiFilters';
+import { useUrlParams } from '../../../hooks/useUrlParams';
+import { history } from '../../../utils/history';
+import { fromQuery, toQuery } from '../Links/url_helpers';
+
+const anyEnvironmentValue = 'All';
+const nullEnvironmentValue = 'Null';
+
+function updateEnvironmentUrl(
+  location: ReturnType<typeof useLocation>,
+  environment?: string
+) {
+  const nextEnvironmentQueryParam =
+    environment !== anyEnvironmentValue ? environment : undefined;
+  history.push({
+    ...location,
+    search: fromQuery({
+      ...toQuery(location.search),
+      environment: nextEnvironmentQueryParam
+    })
+  });
+}
+
+const OptionInputContainer = styled.div`
+  min-width: 128px;
+`;
+
+const EnvironmentFilterLabel = styled.div`
+  height: 100%;
+  border-width: 1px 0 1px 1px;
+  border-style: solid;
+  border-color: #e1e1e1;
+  background: #eee;
+  font-size: ${fontSizes.small};
+  padding: ${px(units.minus)};
+`;
+
+export const EnvironmentFilter: React.FC = () => {
+  const location = useLocation();
+  const { urlParams } = useUrlParams();
+  const { uiFilters } = useUiFilters();
+  const { start, end } = urlParams;
+  const { environment } = uiFilters;
+  const { data = [], status = 'loading' } = useFetcher(
+    () => {
+      if (start && end) {
+        return loadServiceEnvironments({ start, end });
+      }
+    },
+    [start, end]
+  );
+
+  const availableEnvironments = [
+    {
+      value: anyEnvironmentValue,
+      label: i18n.translate('xpack.apm.filter.environment.allLabel', {
+        defaultMessage: 'All'
+      })
+    },
+    ...data.map(environment => ({ value: environment, label: environment })),
+    {
+      value: nullEnvironmentValue,
+      label: i18n.translate('xpack.apm.filter.environment.undefinedLabel', {
+        defaultMessage: 'Undefined'
+      })
+    }
+  ];
+
+  if (
+    environment &&
+    environment !== anyEnvironmentValue &&
+    environment !== nullEnvironmentValue &&
+    status === 'loading'
+  ) {
+    availableEnvironments.push({ value: environment, label: environment });
+  }
+
+  const [valueOfSelected, setValueOfSelected] = useState(
+    environment || anyEnvironmentValue
+  );
+
+  useEffect(
+    () => {
+      setValueOfSelected(environment || anyEnvironmentValue);
+    },
+    [environment]
+  );
```

Comment:
I still don't understand why the selected environment needs to be kept in local state.

Reply:
you're right! i guess it's just a habit for me that input elements have their own state, but in this case, `uiFilters` has the persistence to accomplish the same thing. :D